File: exam-system-java/server/src/main/java/com/yyjy/exam/question/service/QuestionService.java
package com.yyjy.exam.question.service;

import com.alibaba.excel.EasyExcel;
import com.alibaba.excel.write.style.column.LongestMatchColumnWidthStyleStrategy;
import com.yyjy.exam.common.constant.CacheConstant;
import com.yyjy.exam.common.constant.MessageConstant;
import com.yyjy.exam.common.exception.BusinessException;
import com.yyjy.exam.common.util.RedisUtil;
import com.yyjy.exam.entity.question.dto.*;
import com.yyjy.exam.entity.question.entity.QuestionType;
import com.yyjy.exam.entity.question.entity.Questions;
import com.yyjy.exam.entity.question.entity.QuestionsCategoriesTable;
import com.yyjy.exam.entity.question.entity.QuestionsTable;
import com.yyjy.exam.entity.question.io.req.*;
import com.yyjy.exam.entity.question.io.res.QuestionGenerateDto;
import com.yyjy.exam.entity.question.io.res.QuestionPageRes;
import com.yyjy.exam.question.bo.QuestionExcelTemplateBo;
import com.yyjy.exam.question.common.QuestionExcelListener;
import com.yyjy.exam.question.repository.QuestionChoicesRepository;
import com.yyjy.exam.question.repository.QuestionsRepository;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.ast.mutation.AssociatedSaveMode;
import org.babyfish.jimmer.sql.ast.mutation.SaveMode;
import org.springframework.data.redis.core.DefaultTypedTuple;
import org.springframework.data.redis.core.ZSetOperations;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class QuestionService {
	
	private final QuestionsRepository questionsRepository;
	private final QuestionChoicesRepository questionChoicesRepository;
	private final RedisUtil redisUtil;
	private final QuestionImportService questionImportService;
	private final QuestionGenerateService questionGenerateService;
	private final JSqlClient sqlClient;
	
	@Transactional
	public Questions save(QuestionSaveReq req) {
		if (questionsRepository.existsByTypeAndTitle(req.getType(), req.getTitle())) {
			throw new BusinessException(MessageConstant.QUESTION_EXIST);
		}
		
		if (!questionsRepository.existsById(req.getCategoryId())) {
			throw new BusinessException(MessageConstant.QUESTION_CATEGORY_NOT_EXIST);
		}
		
		var builder = new QuestionSaveInput.Builder()
				              .title(req.getTitle())
				              .type(req.getType())
				              .categoryId(req.getCategoryId())
				              .difficulty(req.getDifficulty())
				              .score(req.getScore())
				              .analysis(req.getAnalysis());
		
		if (req.getExtra() instanceof SingleChoiceExtra sc) {
			processChoice(builder, sc.getChoices(), false);
		} else if (req.getExtra() instanceof MultipleChoiceExtra mc) {
			processChoice(builder, mc.getChoices(), true);
		} else if (req.getExtra() instanceof JudgeExtra j) {
			var answers = new QuestionSaveInput.TargetOf_answers();
			answers.setAnswer(j.getAnswer());
			builder.answers(answers);
		} else if (req.getExtra() instanceof TextExtra t) {
			if (t.getAnswer() != null) {
				var answers = new QuestionSaveInput.TargetOf_answers();
				answers.setAnswer(t.getAnswer());
				builder.answers(answers);
			}
		}
		
		return questionsRepository.save(builder.build(), SaveMode.INSERT_ONLY, AssociatedSaveMode.APPEND);
	}
	
	private void processChoice(QuestionSaveInput.Builder builder, List<ChoiceDto> choiceDtos, boolean allowMultiple) {
		if (choiceDtos == null || choiceDtos.isEmpty()) {
			throw new BusinessException(MessageConstant.QUESTION_CHOICE_CHOICES_NOT_EMPTY);
		}
		List<QuestionSaveInput.TargetOf_choices> choices = new ArrayList<>();
		StringBuilder answerStr = new StringBuilder();
		for (int i = 0; i < choiceDtos.size(); i++) {
			ChoiceDto dto = choiceDtos.get(i);
			var choice = new QuestionSaveInput.TargetOf_choices();
			choice.setContent(dto.getContent());
			choice.setCorrect(dto.getCorrect());
			choice.setSort(i);
			choices.add(choice);
			
			if (Boolean.TRUE.equals(dto.getCorrect())) {
				if (!answerStr.isEmpty()) {
					if (!allowMultiple) {
						throw new BusinessException("单选题只能有一个正确答案");
					}
					answerStr.append(",");
				}
				answerStr.append((char) ('A' + i));
			}
		}
		builder.choices(choices);
		
		var answers = new QuestionSaveInput.TargetOf_answers();
		answers.setAnswer(answerStr.toString());
		builder.answers(answers);
	}
	
	@Transactional
	public Questions update(QuestionUpdateInput question) {
		if (!questionsRepository.existsById(question.getCategoryId())) {
			throw new BusinessException(MessageConstant.QUESTION_CATEGORY_NOT_EXIST);
		}
		// 查询题目是否存在
		Questions oldQuestion = questionsRepository.findById(question.getId())
				                        .orElseThrow(() -> new BusinessException(MessageConstant.QUESTION_NOT_EXIST));
		
		// 查询同类型下是否有同名题目
		if (questionsRepository.existsByTypeAndTitle(oldQuestion.type(), question.getTitle())) {
			throw new BusinessException(MessageConstant.QUESTION_EXIST);
		}
		
		// 若是单选或多选题，则删除旧题目选项
		if (QuestionType.SINGLE_CHOICE.equals(oldQuestion.type()) ||
				    QuestionType.MULTIPLE_CHOICE.equals(oldQuestion.type())) {
			questionChoicesRepository.deleteByQuestionId(question.getId());
		}
		
		return questionsRepository.save(question, SaveMode.UPDATE_ONLY);
	}
	
	public void remove(long id) {
		questionsRepository.deleteById(id);
	}
	
	public QuestionsPageView getById(long id) {
		QuestionsPageView question = questionsRepository.sql().getEntities().findById(QuestionsPageView.class, id);
		Optional.ofNullable(question).orElseThrow(() -> new BusinessException(MessageConstant.QUESTION_NOT_EXIST));
		
		CompletableFuture.runAsync(() ->
				                           redisUtil.zIncrementScore(CacheConstant.POPULAR_QUESTIONS_KEY, String.valueOf(id), 1.0)
		);
		
		return question;
	}
	
	public QuestionPageRes list(QuestionListReq req) {
		int page = req.page() != null ? req.page() : 1;
		int size = req.size() != null && req.size() < 20 ? req.size() : 20;
		
		QuestionsTable t = QuestionsTable.$;
		var query = sqlClient.createQuery(t);
		
		if (req.categoryId() != null) {
			query = query.where(t.categoryId().eq(req.categoryId()));
		}
		if (req.difficulty() != null) {
			query = query.where(t.difficulty().eq(req.difficulty()));
		}
		if (req.type() != null) {
			query = query.where(t.type().eq(req.type()));
		}
		var pageResult = query.select(t.fetch(QuestionsPageView.class)).fetchPage(page - 1, size);
		
		List<QuestionsPageView> records = pageResult.getRows();
		return QuestionPageRes.builder()
				       .records(records)
				       .total(pageResult.getTotalRowCount())
				       .pages(pageResult.getTotalPageCount())
				       .size(size)
				       .current(page)
				       .build();
	}
	
	public List<QuestionsPageView> getPopularQuestions(int size) {
		// 从Redis中获取热门题目ID
		Set<ZSetOperations.TypedTuple<String>> popularIdSet =
				redisUtil.zReverseRangeWithScores(CacheConstant.POPULAR_QUESTIONS_KEY, 0, size - 1);
		
		if (popularIdSet == null || popularIdSet.isEmpty()) {
			return questionsRepository.getLast(size);
		}
		
		// 从缓存中获取的题目ID列表
		List<Long> popularIdsList = popularIdSet.stream()
				                            .map(item -> item.getValue() != null ? Long.parseLong(item.getValue()) : null)
				                            .filter(Objects::nonNull)
				                            .toList();
		
		// 从数据库中获取的题目列表
		List<QuestionsPageView> popularQuestionList = new ArrayList<>();
		List<Long> newPopularIdsList = new ArrayList<>();
		
		for (Long questionId : popularIdsList) {
			QuestionsPageView view = questionsRepository.findQuestionsPageViewById(questionId);
			if (view != null) {
				newPopularIdsList.add(questionId);
				popularQuestionList.add(view);
			}
		}
		
		if (popularQuestionList.size() < size) {
			List<QuestionsPageView> notInQuestions = questionsRepository.getLastNotIn(
					size - popularQuestionList.size(), newPopularIdsList);
			popularQuestionList.addAll(notInQuestions);
		}
		
		return popularQuestionList;
	}
	
	/**
	 * 刷新热门题目缓存
	 */
	public void refreshPopularQuestions() {
		// 删除缓存
		redisUtil.delete(CacheConstant.POPULAR_QUESTIONS_KEY);
		// 查询所有题目id
		List<Long> questionIds = questionsRepository.findAllIds();
		// 存入缓存
		Set<ZSetOperations.TypedTuple<String>> tuples = questionIds.stream()
				                                                .map(id -> (ZSetOperations.TypedTuple<String>) new DefaultTypedTuple<>(String.valueOf(id), 0.0))
				                                                .collect(Collectors.toSet());
		redisUtil.zAdd(CacheConstant.POPULAR_QUESTIONS_KEY, tuples);
	}
	
	public void exportTemplate(HttpServletResponse response) {
		try {
			// 设置响应结果类型
			// ms-excel: 微软的excel格式
			response.setContentType("application/vnd.ms-excel");
			response.setCharacterEncoding("utf-8");
			// 防止中文乱码
			String fileName = URLEncoder.encode("Excel题目模板", StandardCharsets.UTF_8);
			// 设置响应头Content-Disposition标识任何格式都以下载方式打开
			response.addHeader("Content-Disposition", "attachment;filename=" + fileName + ".xlsx");
			
			// 生成模板文件
			EasyExcel.write(response.getOutputStream(), QuestionExcelTemplateBo.class)
					.registerWriteHandler(new LongestMatchColumnWidthStyleStrategy())
					.sheet("题目模板")
					.doWrite(List.of(createTemplateBo()));
		} catch (Exception e) {
			throw new BusinessException("导出模板失败");
		}
	}
	
	public List<QuestionImportView> parseExcel(MultipartFile file) {
		try {
			QuestionExcelListener listener = new QuestionExcelListener();
			EasyExcel.read(file.getInputStream(), QuestionExcelTemplateBo.class, listener).sheet().doRead();
			
			List<String> errors = listener.getErrors();
			if (!errors.isEmpty()) {
				throw new BusinessException("数据校验失败：\n" + String.join("\n", errors));
			}
			return listener.getDatas();
		} catch (BusinessException e) {
			throw e;
		} catch (Exception e) {
			throw new BusinessException("解析Excel失败");
		}
	}
	
	public String importBatch(List<QuestionImportInput> questions) {
		return questionImportService.importBatch(questions);
	}
	
	public String importBatchExcel(MultipartFile file) {
		if (file.isEmpty()) {
			throw new BusinessException(MessageConstant.FILE_NOT_EMPTY);
		}
		String filename = file.getOriginalFilename();
		if (filename == null || (!filename.endsWith(".xlsx") && !filename.endsWith(".xls"))) {
			throw new BusinessException(MessageConstant.FILE_FORMAT_ERROR);
		}
		List<QuestionImportView> views = parseExcel(file);
		List<QuestionImportInput> inputs = views.stream()
				                                   .map(this::convertImportViewToInput)
				                                   .toList();
		return questionImportService.importBatch(inputs);
	}
	
	public List<QuestionGenerateDto> generateQuestions(QuestionGenerateReq req) {
		var ct = QuestionsCategoriesTable.$;
		String categoryDb = sqlClient.createQuery(ct)
				                    .where(ct.id().eq(req.categoryId()))
				                    .select(ct.name())
				                    .fetchFirstOrNull();
		
		if (categoryDb == null) {
			throw new BusinessException(MessageConstant.QUESTION_CATEGORY_NOT_EXIST);
		}
		
		return questionGenerateService.generateQuestions(
				categoryDb,
				req.count(),
				req.difficulty(),
				req.types(),
				req.includeMultiple()
		);
	}
	
	private QuestionImportInput convertImportViewToInput(QuestionImportView view) {
		var input = new QuestionImportInput.Builder()
				            .title(view.getTitle())
				            .type(view.getType())
				            .categoryId(view.getCategoryId())
				            .difficulty(view.getDifficulty())
				            .score(view.getScore())
				            .analysis(view.getAnalysis())
				            .answer(view.getAnswer());
		if (view.getChoices() != null) {
			List<QuestionImportInput.TargetOf_choices> choices = new ArrayList<>();
			for (QuestionImportView.TargetOf_choices vChoice : view.getChoices()) {
				QuestionImportInput.TargetOf_choices choice = new QuestionImportInput.TargetOf_choices();
				choice.setContent(vChoice.getContent());
				choice.setCorrect(vChoice.getCorrect());
				choice.setSort(vChoice.getSort());
				choices.add(choice);
			}
			input.choices(choices);
		}
		return input.build();
	}
	
	private QuestionExcelTemplateBo createTemplateBo() {
		return QuestionExcelTemplateBo.builder()
				       .content("以下哪个是Spring框架的核心特性?")
				       .type("CHOICE")
				       .multiple("否")
				       .categoryId("1")
				       .difficulty("MEDIUM")
				       .score("1")
				       .choiceA("依赖注入")
				       .choiceB("面向切面编程")
				       .choiceC("事务管理")
				       .choiceD("以上都是")
				       .answer("D")
				       .analysis("Spring框架的核心特性包括依赖注入、面向切面编程和事务管理等。")
				       .build();
		
	}
}

File: exam-system-server/server/src/main/kotlin/com/yyjy/service/QuestionService.kt
package com.yyjy.service

import cn.hutool.core.collection.CollUtil
import com.alibaba.excel.EasyExcel
import com.alibaba.excel.write.style.column.LongestMatchColumnWidthStyleStrategy
import com.yyjy.common.BusinessException
import com.yyjy.constants.CacheConstant
import com.yyjy.constants.MessageConstant
import com.yyjy.constants.QuestionConstant
import com.yyjy.models.bo.QuestionExcelTemplateBo
import com.yyjy.models.dto.QuestionGenerateDto
import com.yyjy.models.entity.*
import com.yyjy.models.entity.dto.*
import com.yyjy.models.req.QuestionGenerateReq
import com.yyjy.models.req.QuestionListReq
import com.yyjy.models.res.QuestionPageRes
import com.yyjy.repository.QuestionChoicesRepository
import com.yyjy.repository.QuestionsRepository
import com.yyjy.utils.ExcelUtil
import com.yyjy.utils.RedisUtil
import jakarta.servlet.http.HttpServletResponse
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.babyfish.jimmer.sql.ast.mutation.AssociatedSaveMode
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.babyfish.jimmer.sql.kt.ast.expression.eq
import org.babyfish.jimmer.sql.kt.ast.expression.`eq?`
import org.babyfish.jimmer.sql.kt.ast.expression.like
import org.babyfish.jimmer.sql.kt.exists
import org.springframework.data.redis.core.DefaultTypedTuple
import org.springframework.data.redis.core.ZSetOperations
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.multipart.MultipartFile
import java.net.URLEncoder
import kotlin.jvm.optionals.getOrElse

@Service
class QuestionService(
    private val questionsRepository: QuestionsRepository,
    private val questionChoicesRepository: QuestionChoicesRepository,
    private val redisUtil: RedisUtil,
    private val questionImportService: QuestionImportService,
    private val questionGenerateService: QuestionGenerateService
) {

    @Transactional
    fun save(question: QuestionSaveInput): Questions {
        if (questionsRepository.existsByTypeAndTitle(question.type, question.title)) {
            throw BusinessException(MessageConstant.QUESTION_EXIST)
        }
        if (!questionsRepository.sql.exists(Categories::class) {
                where(table.id eq question.categoryId)
            }) {
            throw BusinessException(MessageConstant.QUESTION_CATEGORY_NOT_EXIST)
        }

        if (question.type == "CHOICE") {
            if (question.choices == null) {
                throw BusinessException(MessageConstant.QUESTION_CHOICE_CHOICES_NOT_EMPTY)
            }
            var answerStr = ""
            for ((index, targetofChoices) in question.choices!!.withIndex()) {
                targetofChoices.sort = index
                if (targetofChoices.correct == true) {
                    if (answerStr.isNotEmpty()) {
                        if (question.multi == false) {
                            throw BusinessException(MessageConstant.QUESTION_CHOICE_MULTI_ANSWER_NOT_EMPTY)
                        }
                        answerStr += ","
                    }
                    answerStr += 'A' + index
                }
            }
            question.answers?.answer = answerStr
        } else {
            if (question.multi == true) {
                throw BusinessException(MessageConstant.QUESTION_CHOICE_MULTI_NOT_EMPTY)
            }
        }

        return questionsRepository.save(question) {
            // 1. 解决 <root> 问题：告诉 Jimmer 根对象直接插入，不要去找 ID
            setMode(SaveMode.INSERT_ONLY)
            // 2. 解决 <root>.questionAnswers 问题：告诉 Jimmer 子对象直接追加
            setAssociatedMode(Questions::questionAnswers, AssociatedSaveMode.APPEND)
            // 3. 解决 <root>.questionChoices 问题：告诉 Jimmer 子对象直接追加
            setAssociatedMode(Questions::questionChoices, AssociatedSaveMode.APPEND)
        }
    }

    @Transactional
    fun update(question: QuestionUpdateInput): Questions {
        if (!questionsRepository.sql.exists(Categories::class) {
                where(table.id eq question.categoryId)
            }) {
            throw BusinessException(MessageConstant.QUESTION_CATEGORY_NOT_EXIST)
        }
        val oldQuestion =
            questionsRepository.findById(question.id)
                .getOrElse { throw BusinessException(MessageConstant.QUESTION_NOT_EXIST) }

        // 查询同类型下是否有同名题目
        if (questionsRepository.existsByTypeAndTitle(oldQuestion.type, question.title)) {
            throw BusinessException(MessageConstant.QUESTION_EXIST)
        }

        if (oldQuestion.type == QuestionConstant.TYPE.CHOICE) {
            // 删除所有选择题选项
            questionChoicesRepository.deleteByQuestionId(question.id)
        }

        return questionsRepository.save(question, SaveMode.UPDATE_ONLY)
    }

    fun remove(id: Long) {
        questionsRepository.deleteById(id)
    }

    fun getById(id: Long): QuestionsPageView {
        val question = questionsRepository.sql.entities.findById(QuestionsPageView::class, id)
            ?: throw BusinessException(MessageConstant.QUESTION_NOT_EXIST)
        // 开启一个异步协程 存入缓存
        CoroutineScope(Dispatchers.IO).launch {
            redisUtil.zIncrementScore(CacheConstant.POPULAR_QUESTIONS_KEY, id.toString(), 1.0)
        }
        return question
    }

    fun list(req: QuestionListReq): QuestionPageRes {
        val page = req.page ?: 1
        val size = req.size ?: 10
        val pageResult = questionsRepository.sql.createQuery(Questions::class) {
            where(table.categoryId `eq?` req.categoryId)
            where(table.difficulty `eq?` req.difficulty)
            where(table.type `eq?` req.type)
            req.keyword?.takeIf { it.isNotBlank() }?.let {
                where(table.title like "%${req.keyword}%")
            }
            select(table.fetch(QuestionsPageView::class))
        }.fetchPage(page - 1, size)
        return QuestionPageRes(
            size = size,
            records = pageResult.rows,
            total = pageResult.totalRowCount,
            current = page,
            pages = pageResult.totalPageCount
        )
    }

    fun getPopularQuestions(size: Int): List<QuestionsPageView> {
        val popularIdSet =
            redisUtil.zReverseRangeWithScores(CacheConstant.POPULAR_QUESTIONS_KEY, 0, (size - 1).toLong())
        if (CollUtil.isEmpty(popularIdSet)) {
            return questionsRepository.getLast(size)
        }
        // 从缓存中获取的题目ID列表
        val popularIdsList = popularIdSet?.mapNotNull { item ->
            item.value?.toString()?.toLongOrNull()
        } ?: emptyList()

        // 从数据库中获取的题目列表
        val popularQuestionList = mutableListOf<QuestionsPageView>()

        val newPopularIdsList = mutableListOf<Long>()
        for (questionId in popularIdsList) {
            questionsRepository.findQuestionsPageViewById(questionId)?.let {
                newPopularIdsList.add(questionId)
                popularQuestionList.add(it)
            }
        }

        if (popularQuestionList.size < size) {
            val notInQuestions = questionsRepository.getLastNotIn(size - newPopularIdsList.size, newPopularIdsList)
            popularQuestionList.addAll(notInQuestions)
        }
        return popularQuestionList
    }

    fun refreshPopularQuestions() {
        // 删除缓存
        redisUtil.delete(CacheConstant.POPULAR_QUESTIONS_KEY)
        // 查询所有题目id
        val questionIds = questionsRepository.findAllIds()
        // 存入缓存
        val tuples: Set<ZSetOperations.TypedTuple<Any>> =
            questionIds.mapTo(HashSet()) { DefaultTypedTuple(it.toString(), 0.0) }
        redisUtil.zAdd(CacheConstant.POPULAR_QUESTIONS_KEY, tuples)
    }

    fun exportTemplate(response: HttpServletResponse) {
        // 设置响应结果类型
        // ms-excel: 微软的excel格式
        response.contentType = "application/vnd.ms-excel"
        response.characterEncoding = "utf-8"

        // 防止中文乱码
        val fileName = URLEncoder.encode("Excel题目模板", "UTF-8")
        // 设置响应头Content-Disposition标识任何格式都以下载方式打开
        response.addHeader("Content-Disposition", "attachment;filename=$fileName.xlsx")

        // 生成模板文件
        EasyExcel.write(response.outputStream, QuestionExcelTemplateBo::class.java).registerWriteHandler(
            LongestMatchColumnWidthStyleStrategy()
        ).sheet("题目模板")
            .doWrite(
                listOf(
                    QuestionExcelTemplateBo(
                        content = "以下哪个是Spring框架的核心特性?",
                        type = "CHOICE",
                        multiple = "否",
                        categoryId = "1",
                        difficulty = "MEDIUM",
                        score = "5",
                        choiceA = "依赖注入",
                        choiceB = "面向切面编程",
                        choiceC = "事务管理",
                        choiceD = "以上都是",
                        answer = "D",
                        analysis = "Spring框架的核心特性包括依赖注入、面向切面编程和事务管理等。"
                    )
                )
            )
    }

    fun parseExcel(file: MultipartFile): List<QuestionImportView> {
        return ExcelUtil.parseExcelToQuestions(file)
    }

    fun importBatch(question: List<QuestionImportInput>): String {
       return questionImportService.importBatch(question)
    }

    fun importBatchExcel(file: MultipartFile): String {
        if (file.isEmpty) {
            throw BusinessException(MessageConstant.FILE_NOT_EXIST)
        }
        val filename = file.originalFilename!!
        if (!filename.endsWith(".xlsx") && !filename.endsWith(".xls")) {
            throw BusinessException(MessageConstant.FILE_FORMAT_ERROR)
        }
        // 解析文件
        val questions = ExcelUtil.parseExcelToQuestions(file)
        return questionImportService.importBatch(questions.map {
            QuestionImportInput(it.toEntity())
        })
    }

    fun generateQuestions(req: QuestionGenerateReq): List<QuestionGenerateDto>? {
        // 查询分类名
        val categoryDb = questionsRepository.sql.createQuery(Categories::class) {
            where(table.id eq req.categoryId)
            select(table.name)
        }.fetchFirstOrNull()
        if (categoryDb == null) {
            throw BusinessException(MessageConstant.QUESTION_CATEGORY_NOT_EXIST)
        }

        return questionGenerateService.generateQuestions(
            category = categoryDb,
            count = req.count,
            difficulty = req.difficulty,
            type = req.types,
            includeMultiple = req.includeMultiple
        )
    }
}
